torchtext/experimental/datasets/translation.py

```diff
@@ -0,0 +1,444 @@
+import torch
+import logging
+from torchtext.utils import download_from_url, extract_archive
+from torchtext.vocab import build_vocab_from_iterator
+from torchtext.data.utils import get_tokenizer
+from torchtext.vocab import Vocab
+from torchtext.data.functional import read_text_iterator, create_data_from_iterator
+import os
+import io
+import codecs
+import xml.etree.ElementTree as ET
+from collections import defaultdict
+
+URLS = {
+    'Multi30k': [
+        'http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/training.tar.gz',
+        'http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/validation.tar.gz',
+        'http://www.quest.dcs.shef.ac.uk/wmt17_files_mmt/'
+        'mmt_task1_test2016.tar.gz'
+    ],
+    'WMT14':
+    'https://drive.google.com/uc?export=download&id=0B_bZck-ksdkpM25jRUN2X2UxMm8',
+    'IWSLT':
+    'https://wit3.fbk.eu/archive/2016-01//texts/{}/{}/{}.tgz'
+}
+
+
+def _clean_xml_file(f_xml):
+    f_txt = os.path.splitext(f_xml)[0]
+    with codecs.open(f_txt, mode='w', encoding='utf-8') as fd_txt:
+        root = ET.parse(f_xml).getroot()[0]
+        for doc in root.findall('doc'):
+            for e in doc.findall('seg'):
+                fd_txt.write(e.text.strip() + '\n')
+
+
+def _clean_tags_file(f_orig):
+    xml_tags = [
+        '<url', '<keywords', '<talkid', '<description', '<reviewer',
+        '<translator', '<title', '<speaker'
+    ]
+    f_txt = f_orig.replace('.tags', '')
+    with codecs.open(f_txt, mode='w', encoding='utf-8') as fd_txt, \
+            io.open(f_orig, mode='r', encoding='utf-8') as fd_orig:
+        for l in fd_orig:
+            if not any(tag in l for tag in xml_tags):
+                # TODO: Fix utf-8 next line mark
+                #                fd_txt.write(l.strip() + '\n')
+                #                fd_txt.write(l.strip() + u"\u0085")
+                #                fd_txt.write(l.lstrip())
+                fd_txt.write(l.strip() + '\n')
+
+
+def _construct_filenames(filename, languages):
+    filenames = []
+    for lang in languages:
+        filenames.append(filename + "." + lang)
+    return filenames
+
+
+def _construct_filepaths(paths, src_filename, tgt_filename):
+    src_path = None
+    tgt_path = None
+    for p in paths:
+        src_path = p if src_filename in p else src_path
+        tgt_path = p if tgt_filename in p else tgt_path
+    return (src_path, tgt_path)
+
+
+def _setup_datasets(dataset_name,
```

Comment:
Please take a look at some instructions at the end [here](https://github.com/pytorch/text/issues/742#issue-606545827). For the new dataset abstraction, we first provide the access to the raw datasets (`IterableDataset`) and cache out them for map-style datasets together with some transformers.